PRCS/prc-4.md

```diff
@@ -0,0 +1,151 @@
+
+Instead of bridging, this standard allows trading the game assets on base chain via creating and filling sell orders in an orderbook DEX smart contract.
+
+## Specification
+
+Every PRC-4 compliant contract must implement the `IOrderbookDex` interface and events:
+
+```solidity
+/// @notice Facilitates base-chain trading of an asset that is living on a different app-chain.
+/// @dev The contract should never hold any ETH itself.
+interface IOrderbookDex is IERC165 {
+    struct Order {
+        uint256 assetAmount;
+        uint256 pricePerAsset;
+        bool cancelled;
+    }
+
+    event OrderCreated(address indexed seller, uint256 indexed orderId, uint256 assetAmount, uint256 pricePerAsset);
+    event OrderFilled(
+        address indexed seller,
+        uint256 indexed orderId,
+        address indexed buyer,
+        uint256 assetAmount,
+        uint256 pricePerAsset
+    );
+    event OrderCancelled(address indexed seller, uint256 indexed orderId);
+
+    /// @notice Returns the seller's current `orderId` (index that their new sell order will be mapped to).
+    function getSellerOrderId(address seller) external view returns (uint256);
+
+    /// @notice Returns the Order struct information about an order identified by the combination `<seller, orderId>`.
+    function getOrder(address seller, uint256 orderId) external view returns (Order memory);
+
+    /// @notice Creates a sell order with incremental seller-specific `orderId` for the specified `assetAmount` at specified `pricePerAsset`.
+    /// @dev The order information is saved in a nested mapping `seller address -> orderId -> Order`.
+    /// MUST emit `OrderCreated` event.
+    function createSellOrder(uint256 assetAmount, uint256 pricePerAsset) external;
+
+    /// @notice Consecutively fills an array of orders identified by the combination `<seller, orderId>`,
+    /// by providing an exact amount of ETH and requesting a specific minimum amount of asset to receive.
+    /// @dev Transfers portions of msg.value to the orders' sellers according to the price.
+    /// The sum of asset amounts of filled orders MUST be at least `minimumAsset`.
+    /// If msg.value is more than the sum of orders' prices, it MUST refund the excess back to msg.sender.
+    /// An order whose `cancelled` parameter has value `true` MUST NOT be filled.
+    /// MUST change the `assetAmount` parameter for the specified order according to how much of it was filled.
+    /// MUST emit `OrderFilled` event for each order accordingly.
+    function fillOrdersExactEth(uint256 minimumAsset, address payable[] memory sellers, uint256[] memory orderIds)
```

Comment:
Having two arrays instead of one does not pose any threat to hitting the transaction size limit, as realistic scenarios would never come close to the limits (you wouldn't need to be filling 1000 orders at once). I had previously thought about packing the address (20 bytes = uint160) and orderId as uint96 into one uint256, I disregarded that idea because I wanted to keep things simple and it did not really matter storage-wise. But now that you mentioned it, I'm testing for the gas cost difference between these two approaches with **function call**, and I'm seeing around 10% gas cost difference. Do you think it is worth it?

Reply:
If you look at PRC3, we have 2 different IDs (`tokenId` and `userTokenId`). The idea was that `userTokenId` is the wallet-specific one but we can use the global `tokenId` wherever we don't want to bother passing a `<userTokenId, address>` pair

Reply:
Yes but that's because PRC3 is IERC721, so we **had to** have global tokenId to conform with the IERC721 in functions like `tokenURI(uint256 tokenId)`. We don't need that here, it would just be a waste of storage - If we want to have the single id, we can simply construct it by packing seller address and seller-specific order id into one uint256 id.